src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1598,25 +1593,35 @@ void Session::on_integration_failure(const IntegrationException& error)
     }
 }
 
-void Session::on_changesets_integrated(version_type client_version, const SyncProgress& progress)
+void Session::on_changesets_integrated(version_type client_version, const SyncProgress& progress,
+                                       bool changesets_integrated)
 {
     REALM_ASSERT_EX(m_state == Active, m_state);
     REALM_ASSERT_3(progress.download.server_version, >=, m_download_progress.server_version);
-    m_download_progress = progress.download;
     bool upload_progressed = (progress.upload.client_version > m_progress.upload.client_version);
+
+    m_download_progress = progress.download;
     m_progress = progress;
+
     if (upload_progressed) {
         if (progress.upload.client_version > m_last_version_selected_for_upload) {
             if (progress.upload.client_version > m_upload_progress.client_version)
                 m_upload_progress = progress.upload;
             m_last_version_selected_for_upload = progress.upload.client_version;
         }
 
+        notify_upload_progress();
         check_for_upload_completion();
     }
 
-    do_recognize_sync_version(client_version); // Allows upload process to resume
-    check_for_download_completion();           // Throws
+    bool resume_upload = do_recognize_sync_version(client_version); // Allows upload process to resume
+
+    // notify also when final DOWNLOAD received with no changesets
+    bool download_progressed = changesets_integrated || (!upload_progressed && resume_upload);
```

Comment:
Thinking of it a bit more, isn't `bool download_progressed = changesets_integrated` enough? For download, we should have something to report only when we integrate changesets. It's true that a download has implications on upload too (but that's reported separately and we handle it above)

Reply:
I looked at the original code, and to have the same behavior (+ bootstrap bytes) you'd need to always invoke `notify_download_progress` here (on_sync_progress was essentially moved to on_changesets_integrated). I am wondering though, don't we sometimes get the same progress twice (if both notify_download_progress and notify_upload_progress are invoked in this function)?

I'll take a look at those tests.

Reply:
Interesting, so that may explain why i haven't noticed any issues in object-store tests. I haven't found any assertion or a check in sync code, so i've just assumed here that if explicit check for upload_progressed yields false, then it must be download related update. So it might have been corner case in test sync server implementation - DOWNLOAD message without changesets and change in client and server versions in it.

Anyway, duplicated values are essentially already the problem (i've noted it here #7451), and i'd like to ultimately solve it. Here i've made explicit split for upload/download notify, and the only situation where you'll get the same progress twice is when DOWNLOAD has changesets and acknowledged upload. If you think it should be changed here, then the fix is straightforward: omit report_progress for upload and call here notify_download_progress unconditionally. At least until #7451 is fully solved. I've refactored it already to work for object-store tests, but need still some work on sync tests, because they are a bit weird.

Reply:
I think splitting on_sync_progress into on_download_progress/on_upload_progress has made a bunch of these checks more confusing. Before it was unclear if you didn't know that downloads are the only thing that advances progress (but at least you only had to look in one place to figure that out), and now it's confusing all the time (and you have to check in two places). I don't want to keep arguing about this though.

Reply:
what is the conclusion here?


> about notify_download_progress: it is essentially the replacement for old on_sync_progress and it was always called unconditionally along on_changesets_integrated.

It seems we should always call `notify_download_progress`, right?